receiver/kafkareceiver/avro_unmarshaler.go

```diff
@@ -0,0 +1,184 @@
+	bodyField                 = "body"
+	timestampField            = "timestamp"
+	severityTextField         = "severityText"
+	severityNumberField       = "severityNumber"
+)
+
+var (
+	errNoAvroMapping   = errors.New("no avro field mapping provided")
+	errNoAvroSchemaURL = errors.New("no avro url provided")
+)
+
+type avroLogsUnmarshaler struct {
+	deserializer avroDeserializer
+	mapping      map[string]string
+}
+
+func newAVROLogsUnmarshaler() *avroLogsUnmarshaler {
+	return &avroLogsUnmarshaler{}
+}
+
+func (a *avroLogsUnmarshaler) Encoding() string {
+	return avroEncoding
+}
+
+func (a *avroLogsUnmarshaler) Unmarshal(data []byte) (plog.Logs, error) {
+	avroLog, err := a.deserializer.Deserialize(data)
+	if err != nil {
+		return plog.NewLogs(), fmt.Errorf("failed to deserialize avro log: %w", err)
+	}
+
+	return a.mapLog(avroLog), nil
+}
+
+func (a *avroLogsUnmarshaler) mapLog(avroLog map[string]interface{}) plog.Logs {
+	logs := plog.NewLogs()
+	resourceLogs := logs.ResourceLogs().AppendEmpty()
+	resourceAttributes := resourceLogs.Resource().Attributes()
+	scopeLogs := resourceLogs.ScopeLogs().AppendEmpty()
+	logRecord := scopeLogs.LogRecords().AppendEmpty()
+	logRecordAttributes := logRecord.Attributes()
+
+	for sourceField, target := range a.mapping {
+		switch {
+		case isResourceAttribute(target):
+			targetField := strings.TrimPrefix(target, resourceAttributesPrefix)
+			mapAttribute(avroLog, sourceField, &resourceAttributes, targetField)
+		case isLogRecordAttribute(target):
+			targetField := strings.TrimPrefix(target, logRecordAttributesPrefix)
+			mapAttribute(avroLog, sourceField, &logRecordAttributes, targetField)
+		case target == bodyField:
+			logRecord.Body().SetStr(avroLog[sourceField].(string))
+		case target == timestampField:
+			timestamp := avroLog[timestampField].(time.Time)
+			logRecord.SetTimestamp(pcommon.NewTimestampFromTime(timestamp))
+		case target == severityTextField:
+			logRecord.SetSeverityText(avroLog[sourceField].(string))
+		case target == severityNumberField:
+			severityNumber, ok := avroLog[sourceField].(int32)
+			if !ok {
+				severityNumber = 0
+			}
+			logRecord.SetSeverityNumber(plog.SeverityNumber(severityNumber))
+		}
+	}
+
+	return logs
+}
+
+func isResourceAttribute(targetField string) bool {
+	return strings.HasPrefix(targetField, resourceAttributesPrefix)
+}
+
+func isLogRecordAttribute(targetField string) bool {
+	return strings.HasPrefix(targetField, logRecordAttributesPrefix)
+}
+
+func mapAttribute(avroLog map[string]interface{}, sourceField string, attributes *pcommon.Map, targetField string) {
+	sourceValue := avroLog[sourceField]
+	if sourceValue != nil {
+		setAttribute(attributes, targetField, sourceValue)
+	}
+}
+
+func setAttribute(attributes *pcommon.Map, key string, value interface{}) {
+	attribute := attributes.PutEmpty(key)
+	_ = attribute.FromRaw(value)
+}
+
+func (a *avroLogsUnmarshaler) Init(schemaURL string, mapping map[string]string) error {
```

Comment:
Does the anyone other `Unmarshaler` implement an `Init` method?

I would prefer not to have an `Init` method since it implies the state is mutable and with that requires some additional locking to ensure it is concurrency safe.

Having a constructor method means we can create an immutable serialiser which can be lock free and giving some performance gains.

Reply:
The existing unmarshalers do not use `Init` methods b/c they don't require additional configuration like the mapping. I had no access to the config in `defaultLogsUnmarshalers()` where the unmarshalers are created so I introduced it as a way to pass the config into it.